In [1]:
#to install pandas
!pip install xlrd

twisted 18.7.0 requires PyHamcrest>=1.9.0, which is not installed.
You are using pip version 10.0.1, however version 20.2.3 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [ ]:
#to read exel files
!pip install xlrd

Firstly you need libraries, you can use **Pandas** for data analisys and **numpy** as a fundamental package in **python**. 

In [2]:
import pandas as pd
import numpy as np

How can you create a dataframe?

In [ ]:
# From a list
d = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
df3 = pd.DataFrame(d)
df3


In [ ]:
# From a dictionary
d2 = {'col1': [1, 2], 'col2': [3, 4]}
df2 = pd.DataFrame(d)
df2

In [ ]:
# From an array
d3 = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
df3 = pd.DataFrame(d3, columns=['a', 'b', 'c'])
df3

Usually, a dataframe is created from dictionary. You can write the dictionary directly in the argument of the dataframe.

In [3]:
df = pd.DataFrame({'a':[1,2,1],'b':[1,1,1]})
df

a  b
0  1  1
1  2  1
2  1  1

In [ ]:
df['a']==1

In [ ]:
help(df.to_csv)

## Opening files

The next functions allow us to open files from any place of our computer, or even from a link of internet.

In [23]:
# from a link
csv_path = 'https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/PY0101EN/Chapter%204/Datasets/TopSellingAlbums.csv'
df = pd.read_csv(csv_path)

In [5]:
df.head()

Artist                      Album  Released   Length  \
0  Michael Jackson                   Thriller      1982  0:42:19   
1            AC/DC              Back in Black      1980  0:42:11   
2       Pink Floyd  The Dark Side of the Moon      1973  0:42:49   
3  Whitney Houston              The Bodyguard      1992  0:57:44   
4        Meat Loaf            Bat Out of Hell      1977  0:46:33   

                         Genre  Music Recording Sales (millions)  \
0               pop, rock, R&B                              46.0   
1                    hard rock                              26.1   
2             progressive rock                              24.2   
3               R&B, soul, pop                              27.4   
4  hard rock, progressive rock                              20.6   

   Claimed Sales (millions) Released.1 Soundtrack  Rating  
0                        65  30-Nov-82        NaN    10.0  
1                        50  25-Jul-80        NaN     9.5  
2                        45  01-Mar-73        NaN     9.0  
3                        44  17-Nov-92          Y     8.5  
4                        43  21-Oct-77        NaN     8.0

You can also open from the same files folder where your notebook is situated. For example, an excel file:

In [6]:
xlsx_path = 'TopSellingAlbums.xlsx'

df = pd.read_excel(xlsx_path)
df.head()

Artist                      Album  Released    Length  \
0  Michael Jackson                   Thriller      1982  00:42:19   
1            AC/DC              Back in Black      1980  00:42:11   
2       Pink Floyd  The Dark Side of the Moon      1973  00:42:49   
3  Whitney Houston              The Bodyguard      1992  00:57:44   
4        Meat Loaf            Bat Out of Hell      1977  00:46:33   

                         Genre  Music Recording Sales (millions)  \
0               pop, rock, R&B                              46.0   
1                    hard rock                              26.1   
2             progressive rock                              24.2   
3               R&B, soul, pop                              27.4   
4  hard rock, progressive rock                              20.6   

   Claimed Sales (millions) Released.1 Soundtrack  Rating  
0                        65 1982-11-30        NaN    10.0  
1                        50 1980-07-25        NaN     9.5  
2                        45 1973-03-01        NaN     9.0  
3                        44 1992-11-17          Y     8.5  
4                        43 1977-10-21        NaN     8.0

To view and access one column of the dataframe, you can compute:

        df[['columna']]
        
this column will also be a dataframe. 

In [7]:
x = df[['Length']]
x

Length
0  00:42:19
1  00:42:11
2  00:42:49
3  00:57:44
4  00:46:33
5  00:43:08
6  01:15:54
7  00:40:01

You can access to several columns in the following way: 

In [ ]:
y = df[['Artist','Length','Genre']]
y

To access the dataframe values, the functions **.iloc** and **.loc** are usefull:
    
        df.iloc
        df.loc
       

In [9]:
df.iloc[0, 0]

'Michael Jackson'

In [10]:
df.iloc[3,4]

'R&B, soul, pop'

In [11]:
df.loc[1, 'Released']

1980

You can create little dataframes from the original:

In [13]:
df.loc[0:2, 'Artist':'Released']

Artist                      Album  Released
0  Michael Jackson                   Thriller      1982
1            AC/DC              Back in Black      1980
2       Pink Floyd  The Dark Side of the Moon      1973

Here, you can also use all the previously functions we have seen, for example:

In [14]:
df.iloc[0:2, 0:3]

Artist          Album  Released
0  Michael Jackson       Thriller      1982
1            AC/DC  Back in Black      1980

We can open unusual files, for example, a **.dat** file:

In [24]:
#For open a .dat file with numpy
physioDat = np.fromfile('NormalSinusRhythm.dat')

In [25]:
#We can convert the array into a data frame
df_physioDat = pd.DataFrame(data=physioDat)
df_physioDat.head()

0
0 -7.232905e+307
1 -9.897362e+162
2 -1.450451e+135
3            NaN
4 -1.484775e+188

If we need anilyze a time series, maybe we will find some misssing values. How to drop all missing values?

We can use the anterior example **df_physiodat** and clean all the missing values with diferent functios, as shown in the following cells:

In [13]:
# Interpolating missing values
df_clean1 = df_physioDat.interpolate()

In [14]:
# Forward-fill Missing Values - Using value of next row to fill the missing value
df_clean2 = df_physioDat.ffill()

In [15]:
# Backfill Missing Values - Using value of previous row to fill the missing value
df_clean3 = df_physioDat.bfill()

In [22]:
df_clean3[0]

0         -7.232905e+307
1         -9.897362e+162
2         -1.450451e+135
3         -1.484775e+188
4         -1.484775e+188
               ...      
4399099   -8.061100e+264
4399100   -3.101601e+231
4399101   -8.061100e+264
4399102   -8.061100e+264
4399103   -3.101601e+231
Name: 0, Length: 4399104, dtype: float64

Usually, we need change the structure of dataframe, so the next three cell shows an example:

In [25]:
# We create a linera panda:
df = pd.DataFrame({'No':[1,2,3,4,5,6,7,8,9]})

In [35]:
df['No']

0    1
1    2
2    3
3    4
4    5
5    6
6    7
7    8
8    9
Name: No, dtype: int64

In [36]:
# To reshape the linear panda
n = 3
new_df = pd.DataFrame(dict(enumerate(df["No"].iloc[i:i+n].values for i in range(0,df.shape[0],n))))

**df_physioDat** is an example of dataframe, with shape (4399104, 1), it means that the dataframe contains only one column. In adition, we can reshape **df_physioDat**:

In [22]:
n = 128
new_physioDat = pd.DataFrame(dict(enumerate(df_physioDat[0].iloc[i:i+n].values for i in range(0,df_physioDat.shape[0],n) )))
new_physioDat.head()


0              1              2              3              4      \
0 -7.232905e+307  1.151801e-274  -2.997501e+19 -7.254850e+307 -8.336417e+298   
1 -9.897362e+162  5.847561e-232 -4.508089e+279 -2.851460e+192 -7.294350e+307   
2 -1.450451e+135 -1.250248e+241 -3.646078e+307 -2.454038e+260 -5.910143e+115   
3            NaN            NaN -1.903786e+125  4.426717e-232 -1.941923e+289   
4 -1.484775e+188 -5.338067e+307 -7.208254e+221 -1.991734e+183 -6.734765e+307   

           5              6              7              8              9      \
0 -8.416145e+211 -1.964581e+307 -4.929630e+173 -1.578016e+212 -8.417908e+307   
1 -1.942395e+289  -5.368431e+09 -1.457114e+308  2.681085e-284 -4.682527e+144   
2 -6.170792e+307 -5.681618e+250 -7.984158e+192  4.830861e-232 -4.521397e+279   
3  -7.459649e+86 -6.166403e+307 -1.941925e+289 -3.919058e+202 -8.981882e+307   
4 -4.515849e+279 -2.146145e+144 -1.067613e+308 -1.943121e+289 -5.072740e+144   

   ...          34358          34359          34360          34361  \
0  ... -4.916139e+245            NaN -1.331803e+241 -7.983343e+293   
1  ... -3.100088e+231 -7.983330e+293            NaN -1.332781e+241   
2  ...            NaN -5.725650e+250 -2.111469e+255            NaN   
3  ... -2.111466e+255            NaN -1.331803e+241 -7.983343e+293   
4  ... -1.331478e+241 -7.983343e+293            NaN -1.333106e+241   

           34362          34363          34364          34365          34366  \
0            NaN -5.725650e+250 -1.876869e+255            NaN -7.221471e+221   
1 -4.369923e+245            NaN -3.100843e+231 -1.487013e+284            NaN   
2 -3.100085e+231 -7.983343e+293            NaN -7.225008e+221 -7.557279e+264   
3            NaN -5.725650e+250 -1.759564e+255            NaN -7.223236e+221   
4 -4.369923e+245            NaN -7.219706e+221 -1.394075e+284            NaN   

           34367  
0 -3.245830e+274  
1 -7.223236e+221  
2            NaN  
3 -3.245830e+274  
4 -3.102356e+231  

[5 rows x 34368 columns]

Indeed, yo can get a column of the dataframe and transform it into a numpy array:

In [21]:
# We can convert a panda to array
a = df['No'].to_numpy()